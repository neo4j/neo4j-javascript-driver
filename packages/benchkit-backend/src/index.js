import express from 'express'
import config from './config'
import neo4j from 'neo4j-driver'
import WorkloadExecutor from './workload.executor'
import WorkloadRouter from './workload.router'

const neo4jUrl = `${config.scheme}://${config.hostname}:${config.boltPort}`
const driver = neo4j.driver(neo4jUrl, neo4j.auth.basic(config.username, config.password), {
  logging: neo4j.logging.console(config.logLevel)
})
const executor = WorkloadExecutor(driver)

const app = express()

app.use(express.json({
  type: 'application/json'
}))
app.use(config.workloadRoute, WorkloadRouter(executor.execute, config.workloadRoute))
app.use((err, _req, res, _) => {
  console.log(err)
  res.status(err.status || 500).end()
})

const server = app.listen(config.backendPort, () => {
  console.log(`index.js:${process.pid}:Listening on ${config.backendPort}`)
})

process.on('SIGTERM', () => {
  console.debug('SIGTERM signal received: closing HTTP server')
  server.close(() => {
    driver.close()
      .catch((error) => console.err('Error closing the driver', error))
  })
})
