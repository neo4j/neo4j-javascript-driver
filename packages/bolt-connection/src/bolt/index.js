/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import handshake from './handshake'
import create from './create'
import _BoltProtocol from './bolt-protocol-v4x3'
import _RawRoutingTable from './routing-table-raw'

export * from './stream-observers'

export const BoltProtocol = _BoltProtocol
export const RawRoutingTable = _RawRoutingTable

export default {
  handshake,
  create
}
