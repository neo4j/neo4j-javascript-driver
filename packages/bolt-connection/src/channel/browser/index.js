/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import HeapBuffer from './browser-buf'
import WebSocketChannel from './browser-channel'
import BrowserHosNameResolver from './browser-host-name-resolver'
import utf8Codec from './browser-utf8'

/*

This module exports a set of components to be used in browser environment.
They are not compatible with NodeJS environment.
All files import/require APIs from `node/index.js` by default.
Such imports are replaced at build time with `browser/index.js` when building a browser bundle.

NOTE: exports in this module should have exactly the same names/structure as exports in `node/index.js`.

 */

export const alloc = arg => new HeapBuffer(arg)
export const Channel = WebSocketChannel
export const HostNameResolver = BrowserHosNameResolver
export const utf8 = utf8Codec
