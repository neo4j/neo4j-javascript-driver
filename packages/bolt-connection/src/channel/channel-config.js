/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { newError, error, internal } from 'neo4j-driver-core'

const {
  util: { ENCRYPTION_OFF, ENCRYPTION_ON }
} = internal

const { SERVICE_UNAVAILABLE } = error

const ALLOWED_VALUES_ENCRYPTED = [
  null,
  undefined,
  true,
  false,
  ENCRYPTION_ON,
  ENCRYPTION_OFF
]

const ALLOWED_VALUES_TRUST = [
  null,
  undefined,
  'TRUST_ALL_CERTIFICATES',
  'TRUST_CUSTOM_CA_SIGNED_CERTIFICATES',
  'TRUST_SYSTEM_CA_SIGNED_CERTIFICATES'
]

export default class ChannelConfig {
  /**
   * @constructor
   * @param {ServerAddress} address the address for the channel to connect to.
   * @param {Object} driverConfig the driver config provided by the user when driver is created.
   * @param {string} connectionErrorCode the default error code to use on connection errors.
   * @param {object} clientCertificate the client certificate
   */
  constructor (address, driverConfig, connectionErrorCode, clientCertificate) {
    this.address = address
    this.encrypted = extractEncrypted(driverConfig)
    this.trust = extractTrust(driverConfig)
    this.trustedCertificates = extractTrustedCertificates(driverConfig)
    this.knownHostsPath = extractKnownHostsPath(driverConfig)
    this.connectionErrorCode = connectionErrorCode || SERVICE_UNAVAILABLE
    this.connectionTimeout = driverConfig.connectionTimeout
    this.clientCertificate = clientCertificate
  }
}

function extractEncrypted (driverConfig) {
  const value = driverConfig.encrypted
  if (ALLOWED_VALUES_ENCRYPTED.indexOf(value) === -1) {
    throw newError(
      `Illegal value of the encrypted setting ${value}. Expected one of ${ALLOWED_VALUES_ENCRYPTED}`
    )
  }
  return value
}

function extractTrust (driverConfig) {
  const value = driverConfig.trust
  if (ALLOWED_VALUES_TRUST.indexOf(value) === -1) {
    throw newError(
      `Illegal value of the trust setting ${value}. Expected one of ${ALLOWED_VALUES_TRUST}`
    )
  }
  return value
}

function extractTrustedCertificates (driverConfig) {
  return driverConfig.trustedCertificates || []
}

function extractKnownHostsPath (driverConfig) {
  return driverConfig.knownHosts || null
}
