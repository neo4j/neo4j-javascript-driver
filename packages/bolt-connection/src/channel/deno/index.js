/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import DenoChannel from './deno-channel'
import DenoHostNameResolver from './deno-host-name-resolver'
import DenoClientCertificatesLoader from './deno-client-certificates-loader'
/*

 This module exports a set of components to be used in deno environment.
 They are not compatible with NodeJS environment.
 All files import/require APIs from `node/index.js` by default.
 Such imports are replaced at build time with `deno/index.js` when building a deno bundle.

 NOTE: exports in this module should have exactly the same names/structure as exports in `node/index.js`.

  */
export const Channel = DenoChannel
export const HostNameResolver = DenoHostNameResolver
export const ClientCertificatesLoader = DenoClientCertificatesLoader
