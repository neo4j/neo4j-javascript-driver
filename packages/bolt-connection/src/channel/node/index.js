/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import NodeBuffer from './node-buf'
import NodeChannel from './node-channel'
import NodeHostNameResolver from './node-host-name-resolver'
import utf8Codec from './node-utf8'

/*

This module exports a set of components to be used in NodeJS environment.
They are not compatible with browser environment.
All files that require environment-dependent APIs should import this file by default.
Imports/requires are replaced at build time with `browser/index.js` when building a browser bundle.

NOTE: exports in this module should have exactly the same names/structure as exports in `browser/index.js`.

 */

export const alloc = arg => new NodeBuffer(arg)
export const Channel = NodeChannel
export const HostNameResolver = NodeHostNameResolver
export const utf8 = utf8Codec
