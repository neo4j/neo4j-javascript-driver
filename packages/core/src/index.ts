/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  newError,
  Neo4jError,
  PROTOCOL_ERROR,
  SERVICE_UNAVAILABLE,
  SESSION_EXPIRED
} from './error'
import Integer, { int, isInt, inSafeRange, toNumber, toString } from './integer'
import {
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time
} from './temporal-types'
import {
  StandardDate,
  NumberOrInteger,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment
} from './graph-types'
import Record from './record'
import { isPoint, Point } from './spatial-types'
import ResultSummary, {
  queryType,
  ServerInfo,
  Notification,
  NotificationPosition,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats
} from './result-summary'
import Result, { QueryResult, ResultObserver } from './result'
import ConnectionProvider from './connection-provider'
import Connection from './connection'
import Transaction from './transaction'
import Session, { TransactionConfig } from './session'
import Driver, * as driver from './driver'
import auth from './auth'
import * as types from './types'
import * as json from './json'
import * as internal from './internal' // todo: removed afterwards

/**
 * Object containing string constants representing predefined {@link Neo4jError} codes.
 */
const error = {
  SERVICE_UNAVAILABLE,
  SESSION_EXPIRED,
  PROTOCOL_ERROR
}

/**
 * @private
 */
const forExport = {
  newError,
  Neo4jError,
  error,
  Integer,
  int,
  isInt,
  inSafeRange,
  toNumber,
  toString,
  internal,
  isPoint,
  Point,
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment,
  Record,
  ResultSummary,
  queryType,
  ServerInfo,
  Notification,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats,
  Result,
  Transaction,
  Session,
  Driver,
  Connection,
  types,
  driver,
  json,
  auth
}

export {
  newError,
  Neo4jError,
  error,
  Integer,
  int,
  isInt,
  inSafeRange,
  toNumber,
  toString,
  internal,
  isPoint,
  Point,
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment,
  Record,
  ResultSummary,
  queryType,
  ServerInfo,
  Notification,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats,
  Result,
  ConnectionProvider,
  Connection,
  Transaction,
  Session,
  Driver,
  types,
  driver,
  json,
  auth
}

export type {
  StandardDate,
  NumberOrInteger,
  NotificationPosition,
  QueryResult,
  ResultObserver,
  TransactionConfig,
};

export default forExport
