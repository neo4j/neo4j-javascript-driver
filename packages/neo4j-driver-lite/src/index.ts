/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import VERSION from './version'
import { logging } from './logging'

import {
  auth,
  AuthTokenManagers,
  authTokenManagers,
  BookmarkManager,
  bookmarkManager,
  BookmarkManagerConfig,
  Connection,
  ConnectionProvider,
  Date,
  DateTime,
  Driver,
  driver as coreDriver,
  Duration,
  EagerResult,
  error,
  inSafeRange,
  int,
  Integer,
  internal,
  isDate,
  isDateTime,
  isDuration,
  isInt,
  isLocalDateTime,
  isLocalTime,
  isNode,
  isPath,
  isPathSegment,
  isPoint,
  isRelationship,
  isRetriableError,
  isTime,
  isUnboundRelationship,
  LocalDateTime,
  LocalTime,
  ManagedTransaction,
  Neo4jError,
  Node,
  Notification,
  notificationCategory,
  NotificationCategory,
  notificationClassification,
  NotificationClassification,
  NotificationFilter,
  NotificationFilterDisabledCategory,
  notificationFilterDisabledCategory,
  NotificationFilterDisabledClassification,
  notificationFilterDisabledClassification,
  AuthTokenManager,
  AuthTokenAndExpiration,
  staticAuthTokenManager,
  NotificationFilterMinimumSeverityLevel,
  notificationFilterMinimumSeverityLevel,
  NotificationPosition,
  notificationSeverityLevel,
  NotificationSeverityLevel,
  Path,
  PathSegment,
  Plan,
  Point,
  ProfiledPlan,
  QueryConfig,
  QueryResult,
  QueryStatistics,
  Record,
  RecordShape,
  Relationship,
  Result,
  ResultObserver,
  ResultSummary,
  ResultTransformer,
  resultTransformers,
  routing,
  RoutingControl,
  ServerInfo,
  Session,
  SessionConfig,
  Time,
  toNumber,
  toString,
  Transaction,
  TransactionPromise,
  types as coreTypes,
  UnboundRelationship,
  ClientCertificate,
  ClientCertificateProvider,
  ClientCertificateProviders,
  RotatingClientCertificateProvider,
  clientCertificateProviders,
  resolveCertificateProvider
} from 'neo4j-driver-core'
import { DirectConnectionProvider, RoutingConnectionProvider } from 'neo4j-driver-bolt-connection'

type AuthToken = coreTypes.AuthToken
type Config = coreTypes.Config
type InternalConfig = coreTypes.InternalConfig
type TrustStrategy = coreTypes.TrustStrategy
type EncryptionLevel = coreTypes.EncryptionLevel
type SessionMode = coreTypes.SessionMode
type Logger = internal.logger.Logger
type ConfiguredCustomResolver = internal.resolver.ConfiguredCustomResolver

const { READ, WRITE } = coreDriver

const {
  util: { ENCRYPTION_ON, assertString, isEmptyObjectOrNull },
  serverAddress: { ServerAddress },
  urlUtil
} = internal

const USER_AGENT = 'neo4j-javascript/' + VERSION

function isAuthTokenManager (value: unknown): value is AuthTokenManager {
  if (typeof value === 'object' &&
    value != null &&
    'getToken' in value &&
    'handleSecurityException' in value) {
    const manager = value as AuthTokenManager

    return typeof manager.getToken === 'function' &&
      typeof manager.handleSecurityException === 'function'
  }

  return false
}

function createAuthManager (authTokenOrProvider: AuthToken | AuthTokenManager): AuthTokenManager {
  if (isAuthTokenManager(authTokenOrProvider)) {
    return authTokenOrProvider
  }

  let authToken: AuthToken = authTokenOrProvider
  // Sanitize authority token. Nicer error from server when a scheme is set.
  authToken = authToken ?? {}
  authToken.scheme = authToken.scheme ?? 'none'
  return staticAuthTokenManager({ authToken })
}

/**
 * Construct a new Neo4j Driver. This is your main entry point for this
 * library.
 *
 * @param {string} url The URL for the Neo4j database, for instance "neo4j://localhost" and/or "bolt://localhost"
 * @param {Map<string,string>| function()} authToken Authentication credentials. See {@link auth} for helpers.
 * @param {Config} config Configuration object. See the configuration section above for details.
 * @returns {Driver}
 */
function driver (
  url: string,
  authToken: AuthToken | AuthTokenManager,
  config: Config = {}
): Driver {
  assertString(url, 'Bolt URL')
  const parsedUrl = urlUtil.parseDatabaseUrl(url)

  // enabling set boltAgent
  const _config = config as unknown as InternalConfig

  // Determine entryption/trust options from the URL.
  let routing = false
  let encrypted = false
  let trust: TrustStrategy | undefined
  switch (parsedUrl.scheme) {
    case 'bolt':
      break
    case 'bolt+s':
      encrypted = true
      trust = 'TRUST_SYSTEM_CA_SIGNED_CERTIFICATES'
      break
    case 'bolt+ssc':
      encrypted = true
      trust = 'TRUST_ALL_CERTIFICATES'
      break
    case 'neo4j':
      routing = true
      break
    case 'neo4j+s':
      encrypted = true
      trust = 'TRUST_SYSTEM_CA_SIGNED_CERTIFICATES'
      routing = true
      break
    case 'neo4j+ssc':
      encrypted = true
      trust = 'TRUST_ALL_CERTIFICATES'
      routing = true
      break
    default:
      throw new Error(`Unknown scheme: ${parsedUrl.scheme ?? 'null'}`)
  }

  // Encryption enabled on URL, propagate trust to the config.
  if (encrypted) {
    // Check for configuration conflict between URL and config.
    if ('encrypted' in _config || 'trust' in _config) {
      throw new Error(
        'Encryption/trust can only be configured either through URL or config, not both'
      )
    }
    _config.encrypted = ENCRYPTION_ON
    _config.trust = trust
    _config.clientCertificate = resolveCertificateProvider(config.clientCertificate)
  }

  const authTokenManager = createAuthManager(authToken)

  // Use default user agent or user agent specified by user.
  _config.userAgent = _config.userAgent ?? USER_AGENT
  _config.boltAgent = internal.boltAgent.fromVersion(VERSION)

  const address = ServerAddress.fromUrl(parsedUrl.hostAndPort)

  const meta = {
    address,
    typename: routing ? 'Routing' : 'Direct',
    routing
  }

  return new Driver(meta, _config, createConnectionProviderFunction())

  function createConnectionProviderFunction (): (id: number, config: Config, log: Logger, hostNameResolver: ConfiguredCustomResolver) => ConnectionProvider {
    if (routing) {
      return (
        id: number,
        config: InternalConfig,
        log: Logger,
        hostNameResolver: ConfiguredCustomResolver
      ): ConnectionProvider =>
        new RoutingConnectionProvider({
          id,
          config,
          log,
          hostNameResolver,
          authTokenManager,
          address,
          userAgent: config.userAgent,
          boltAgent: config.boltAgent,
          routingContext: parsedUrl.query
        })
    } else {
      if (!isEmptyObjectOrNull(parsedUrl.query)) {
        throw new Error(
          `Parameters are not supported with none routed scheme. Given URL: '${url}'`
        )
      }

      return (id: number, config: InternalConfig, log: Logger): ConnectionProvider =>
        new DirectConnectionProvider({
          id,
          config,
          log,
          authTokenManager,
          address,
          userAgent: config.userAgent,
          boltAgent: config.boltAgent
        })
    }
  }
}

/**
 * Verifies if the driver can reach a server at the given url.
 *
 * @experimental
 * @since 5.0.0
 * @param {string} url The URL for the Neo4j database, for instance "neo4j://localhost" and/or "bolt://localhost"
 * @param {Pick<Config, 'logging'>} config Configuration object. See the {@link driver}
 * @returns {true} When the server is reachable
 * @throws {Error} When the server is not reachable or the url is invalid
 */
async function hasReachableServer (url: string, config?: Pick<Config, 'logging'>): Promise<true> {
  const nonLoggedDriver = driver(url, { scheme: 'none', principal: '', credentials: '' }, config)
  try {
    await nonLoggedDriver.getNegotiatedProtocolVersion()
    return true
  } finally {
    await nonLoggedDriver.close()
  }
}

/**
 * Object containing constructors for all neo4j types.
 */
const types = {
  Node,
  Relationship,
  UnboundRelationship,
  PathSegment,
  Path,
  Result,
  EagerResult,
  ResultSummary,
  Record,
  Point,
  Date,
  DateTime,
  Duration,
  LocalDateTime,
  LocalTime,
  Time,
  Integer
}

/**
 * Object containing string constants representing session access modes.
 */
const session = {
  READ,
  WRITE
}

/**
 * Object containing functions to work with {@link Integer} objects.
 */
const integer = {
  toNumber,
  toString,
  inSafeRange
}

/**
 * Object containing functions to work with spatial types, like {@link Point}.
 */
const spatial = {
  isPoint
}

/**
 * Object containing functions to work with temporal types, like {@link Time} or {@link Duration}.
 */
const temporal = {
  isDuration,
  isLocalTime,
  isTime,
  isDate,
  isLocalDateTime,
  isDateTime
}

/**
 * Object containing functions to work with graph types, like {@link Node} or {@link Relationship}.
 */
const graph = {
  isNode,
  isPath,
  isPathSegment,
  isRelationship,
  isUnboundRelationship
}

/**
 * @private
 */
const forExport = {
  authTokenManagers,
  driver,
  hasReachableServer,
  int,
  isInt,
  isPoint,
  isDuration,
  isLocalTime,
  isTime,
  isDate,
  isLocalDateTime,
  isDateTime,
  isNode,
  isPath,
  isPathSegment,
  isRelationship,
  isUnboundRelationship,
  integer,
  Neo4jError,
  isRetriableError,
  auth,
  logging,
  types,
  session,
  routing,
  error,
  graph,
  spatial,
  temporal,
  Driver,
  Result,
  EagerResult,
  Record,
  ResultSummary,
  Node,
  Relationship,
  UnboundRelationship,
  PathSegment,
  Path,
  Integer,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Notification,
  ServerInfo,
  Session,
  Transaction,
  ManagedTransaction,
  TransactionPromise,
  Point,
  Duration,
  LocalTime,
  Time,
  Date,
  LocalDateTime,
  DateTime,
  ConnectionProvider,
  Connection,
  bookmarkManager,
  resultTransformers,
  notificationCategory,
  notificationClassification,
  notificationSeverityLevel,
  notificationFilterDisabledCategory,
  notificationFilterDisabledClassification,
  notificationFilterMinimumSeverityLevel,
  clientCertificateProviders
}

export {
  authTokenManagers,
  driver,
  hasReachableServer,
  int,
  isInt,
  isPoint,
  isDuration,
  isLocalTime,
  isTime,
  isDate,
  isLocalDateTime,
  isDateTime,
  isNode,
  isPath,
  isPathSegment,
  isRelationship,
  isUnboundRelationship,
  integer,
  Neo4jError,
  isRetriableError,
  auth,
  logging,
  types,
  session,
  routing,
  error,
  graph,
  spatial,
  temporal,
  Driver,
  Result,
  EagerResult,
  Record,
  ResultSummary,
  Node,
  Relationship,
  UnboundRelationship,
  PathSegment,
  Path,
  Integer,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Notification,
  ServerInfo,
  Session,
  Transaction,
  ManagedTransaction,
  TransactionPromise,
  Point,
  Duration,
  LocalTime,
  Time,
  Date,
  LocalDateTime,
  DateTime,
  ConnectionProvider,
  Connection,
  bookmarkManager,
  resultTransformers,
  notificationCategory,
  notificationClassification,
  notificationSeverityLevel,
  notificationFilterDisabledCategory,
  notificationFilterDisabledClassification,
  notificationFilterMinimumSeverityLevel,
  clientCertificateProviders
}
export type {
  QueryResult,
  AuthToken,
  AuthTokenManager,
  AuthTokenManagers,
  AuthTokenAndExpiration,
  Config,
  EncryptionLevel,
  TrustStrategy,
  SessionMode,
  ResultObserver,
  NotificationPosition,
  BookmarkManager,
  BookmarkManagerConfig,
  SessionConfig,
  QueryConfig,
  RoutingControl,
  RecordShape,
  ResultTransformer,
  NotificationCategory,
  NotificationClassification,
  NotificationSeverityLevel,
  NotificationFilter,
  NotificationFilterDisabledCategory,
  NotificationFilterDisabledClassification,
  NotificationFilterMinimumSeverityLevel,
  ClientCertificate,
  ClientCertificateProvider,
  ClientCertificateProviders,
  RotatingClientCertificateProvider
}
export default forExport
