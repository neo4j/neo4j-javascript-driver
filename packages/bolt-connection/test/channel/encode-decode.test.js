/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import crypto from 'crypto'
import { utf8 } from '../../src/channel'

const { encode, decode } = utf8

describe('uft8', () => {
  it.each([
    makeString(85000),
    '1234567890',
    '',
    '±!@#$%^&*()_+~`\'|][{}=-+±§<>,."',
    'àáâäæãåā',
    'èéêëēėę',
    'îïíīīįì',
    'ôöòóœøōõ',
    'ûüùúū',
    '⚡ 🏃‍♀️'
  ])('should decode encoded string', str => {
    const encoded = encode(str)
    
    const encodedThenDecoded = decode(encoded, encoded.length)

    expect(encodedThenDecoded).toEqual(str)
  })

  it.each([
    '1234567890',
    '',
    '±!@#$%^&*()_+~`\'|][{}=-+±§<>,."',
    'àáâäæãåā',
    'èéêëēėę',
    'îïíīīįì',
    'ôöòóœøōõ',
    'ûüùúū',
    '⚡ 🏃‍♀️'
  ])('.encode("%s") should match snapshot', str => {
    const encoded = encode(str)
    
    expect(encoded.toString()).toMatchSnapshot()
  })
})


function makeString ( len ) {
  return crypto.randomBytes(len).toString('hex')
}
