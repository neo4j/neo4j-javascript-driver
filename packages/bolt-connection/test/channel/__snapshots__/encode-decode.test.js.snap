// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uft8 .encode("") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  "
`;

exports[`uft8 .encode("1234567890") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  31 32 33 34 35 36 37 38 39 30"
`;

exports[`uft8 .encode("±!@#$%^&*()_+~\`'|][{}=-+±§<>,."") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c2 b1 21 40 23 24 25 5e 26 2a 28 29 5f 2b 7e 60 27 7c 5d 5b 7b 7d 3d 2d 2b c2 b1 c2 a7 3c 3e 2c 2e 22"
`;

exports[`uft8 .encode("àáâäæãåā") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c3 a0 c3 a1 c3 a2 c3 a4 c3 a6 c3 a3 c3 a5 c4 81"
`;

exports[`uft8 .encode("èéêëēėę") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c3 a8 c3 a9 c3 aa c3 ab c4 93 c4 97 c4 99"
`;

exports[`uft8 .encode("îïíīīįì") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c3 ae c3 af c3 ad c4 ab c4 ab c4 af c3 ac"
`;

exports[`uft8 .encode("ôöòóœøōõ") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c3 b4 c3 b6 c3 b2 c3 b3 c5 93 c3 b8 c5 8d c3 b5"
`;

exports[`uft8 .encode("ûüùúū") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  c3 bb c3 bc c3 b9 c3 ba c5 ab"
`;

exports[`uft8 .encode("⚡ 🏃‍♀️") should match snapshot 1`] = `
"ChannelBuffer( position=0 )
  e2 9a a1 20 f0 9f 8f 83 e2 80 8d e2 99 80 ef b8 8f"
`;
